lib/pmdtester/builders/rule_set_builder.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+require 'nokogiri'
+require 'set'
+require_relative '../cmd'
+module PmdTester
+  # This class is responsible for generation dynamic configuration
+  # according to the difference between base and patch branch of Pmd.
+  class RuleSetBuilder
+    ALL_RULE_SETS = Set['bestpractices', 'codestyle', 'design', 'documentation', 'errorprone',
+                        'migrating', 'multithreading', 'performance', 'regex', 'security'].freeze
+    PATH_TO_PMD_JAVA_BASED_RULES =
+      'pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/'
+    PATH_TO_PMD_XPATH_BASED_RULES = 'pmd-java/src/main/resources/category/java/'
+    PATH_TO_ALL_JAVA_RULES = 'config/all-java.xml'
+    PATH_TO_DYNAMIC_CONFIG = 'target/dynamic-config.xml'
+    NO_JAVA_RULES_CHANGED_MESSAGE = 'No java rules have been changed!'
+
+    def initialize(options)
+      @options = options
+    end
+
+    def build
+      filenames = diff_filenames
+      rule_sets = get_rule_sets(filenames)
+      build_config_file(rule_sets)
+    end
+
+    def filter_set(rule_sets)
+      if rule_sets.size.eql?(ALL_RULE_SETS.size)
+        # if `rule_sets` contains all rule sets, than no need to filter the baseline
+        @options.filter_set = nil
+      else
+        @options = rule_sets
+      end
+    end
+
+    def diff_filenames
+      filenames = nil
+      Dir.chdir(@options.local_git_repo) do
+        diff_cmd = "git diff --name-only #{@options.base_branch}..#{@options.patch_branch}"
+        grep_cmd = "grep -E '^pmd-(core|java)/'" # OPTIMIZE: extend other languages
+        filenames = Cmd.execute("#{diff_cmd} | #{grep_cmd}")
+      end
+      filenames.split("\n")
+    end
+
+    def get_rule_sets(filenames)
+      rule_sets = Set[]
+      filenames.each do |filename|
+        if filename.start_with?(PATH_TO_PMD_JAVA_BASED_RULES, PATH_TO_PMD_XPATH_BASED_RULES)
+          add_rule_set(rule_sets, filename)
```

Comment:
Instead of searching in the Set of known categories, you could also use a regex pattern to extract the category directly from the filenames.
There are two cases:
1. filename.start_with?(PATH_TO_PMD_JAVA_BASED_RULES)
`category = %r{pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/([^/]+)/[^/]+Rule.java}.match(filename)[1]`

It defines one capturing group to extract the category from the directory name.

2. filename.start_with?(PATH_TO_PMD_XPATH_BASED_RULES)
`category = %{pmd-java/src/main/resources/category/java/([^/]+).xml}.match(filename)[1]`

This captures the filename without the extension.

See https://ruby-doc.org/core-2.1.1/Regexp.html#class-Regexp-label-Capturing

Note: If there is no match, then `match` returns nil. Keep that in mind, when you eliminate the start_with check - since this can be checked by the regex, too. either it matches (and a match returns a MatchData object) or it doesn't, then you need to do the second regex.

Reply:
We could later on generalize these regex even more, e.g. extracting also the language... but not now.